# Set up:

- **Do NOT change the name of this jupyter notebook! Otherwise, the autograder will not work as intended!**
- **There will be some cells that appear blank and that you can't type in. That is normal. Just ignore them.**
- **Complete your homework on JupyterHub. Do not copy the file to your local machine.**
- Any files you create for this assignment should be stored in the same directory where you found this notebook. You can also upload your log to that directory when you are finished with the assignment.

   *You don't have to do anything here, just don't CHANGE the path where you save or create files when working on JupyterHub!*
- Once you are finished, you can submit your assignment directly through JupyterHub and we will run the autograder and return feedback to you.
- Once your are finished, please **DO** download a copy of your submitted/{assignment name} folder from JupyterHub to your local machine and commit changes and push to your GitHub repository for your records!
---
Before you turn this assignment in, make sure everything runs as expected. First, **restart the kernel** (in the menubar, select Kernel $\rightarrow$ Restart) and then **run all cells** (in the menubar, select Cell $\rightarrow$ Run All). You can also *Validate* your assignment in the Nbgrader Assignment List to make sure it is correct.

Make sure you fill in any place that says `YOUR CODE HERE` and remove the statement `raise NotImplementedError()` below it.

You are certainly allowed to work together and communicate with your peers to solve this assignment but per UCF Academic Integrity policy, **the work in the cells below MUST be your own unique work to receive credit.** 

# Homework 4
---

## Problem 1
**1.1** Recall the quadratic formula
$$x = \frac{-b\pm\sqrt{b^2-4ac}}{2a}$$
which gives the roots of $ax^2 + bx + c = 0$. 
- Write a function called `quadratic` that calculates and returns a tuple containing both roots, given the three coefficients $a$, $b$, and $c$. Be sure to include an informative docstring!
- Make the function arguments ($a$, $b$, and $c$) *keyword arguments* with default values of $0$ so that the user doesn't have to supply one of them if it's value in the expression is $0$.

*In this problem you do not have to worry about complex roots if you do not want to! We will only test expressions with **real** roots.*

In [5]:
import math
def quadratic(a=0,b=0,c=0):
    """
    calculates the equation ax^2+bx+c with coefficients a, b, and c 
    returns a tuple containing roots
    calculates discriminate first
    then take the square root of discriminate by calling math function
    divides all of that by 2a
    inputs:
    coefficents (integers) a , b , c
    outputs:
    tuple with roots 
    """
    discriminate = (b**2) - 4*a*c #inside of sqrt of quadratic formula
    root_1 = (-b + math.sqrt(discriminate)) / (2*a)
    root_2 = (-b - math.sqrt(discriminate)) / (2*a)
    solution = root_1, root_2 # shows result as tuple
    return solution
    
    

**1.2**
- Calculate the roots of $x^2-x-12=0$ and return them as the tuple `tup1`.
- Calculate the roots of $2x^2 - 2x = 0$ and return them as the tuple `tup2`

In [6]:
tup1 = quadratic(1,-1,-12)
tup2 = quadratic(2,-2)
print(tup1, tup2)

(4.0, -3.0) (1.0, 0.0)


##### Expected Call/Output:
```
tup1 = quadratic(1,-1,-12)
tup2 = quadratic(2,-2)
print(tup1, tup2)
```
`(4.0, -3.0) (1.0, 0.0)`


In [4]:
assert tup1[0] == 4 and tup1[1] == -3 and tup2[0] == 1 and tup2[1] == 0
    

## Problem 2
**2.1** Create a *numpy array* called `student_h` that constains lists with the heights of students in feet and inches. A person being 5' 6" would be noted as [5, 6], a person being 5' 10.5" would be noted as [5, 10.5] etc. Populate your array with the heights of five students who are: 5'1", 4'10", 6'2", 5'7", and 6'11". Print it.

In [7]:
import numpy as np
student_h = np.array([[5,1],[4,10],[6,2],[5,7],[6,11]]) # adds values as pairs
print(student_h)


[[ 5  1]
 [ 4 10]
 [ 6  2]
 [ 5  7]
 [ 6 11]]


**2.2** Create a function called `ft_to_m` in the cell below that takes as input `student_h` and returns an array with heights in a the form [m, cm], while always rounding centimeters up. 

Use $1~\mathrm{ft}~=~0.3048~\mathrm{m}$ and $1~\mathrm{inch}~=~0.0254~\mathrm{m}$ and $1~\mathrm{m}~=~100~\mathrm{cm}$.

*You are not expected to use* `for` *loops in this problem but you will need to be clever with array slicing.*

In [20]:
def ft_to_m(student_h):
    """
    takes in different height pairs [feet,inches] and converts into form [m,cm]
    multiples value of feet by meters conversion factor
    multiplies value of inches by meters then centimeters conversion factor
    input:
    array values from student_h [ft,in]
    output:
    heights as array [m,cm]
    """
    feet = student_h[:,0] #defines feet as the first column (0)
    inches = student_h[:,1] #defines inches as the 2nd column (1)
    total_inches = feet * 12 + inches #convert ft into inches so all values same unit
    total_cm = total_inches * 2.54 #convert all inches to cm so all values same unit
    meters = (total_cm//100).astype(int) # takes the whole number integer, divide by 100cm to convert to m
    centimeters = np.ceil(total_cm%100).astype(int) # gives the remainder after finding full number integers 
    new_list = np.array([meters, centimeters]).T
    return new_list

    



**2.3** Call your function for `student_h` and print all students' converted heights in one command. Check that the outputs for each individual student are correct by hand, by using a calculator, or by writing a test case here in python.

In [21]:
call = ft_to_m(student_h)
print(call)

[[ 1 55]
 [ 1 48]
 [ 1 88]
 [ 1 71]
 [ 2 11]]


##### Expected Call/Output:
```
print("The students' heights in [m cm] are:\n", ft_to_m(student_h))
```
```
The students' heights in [m cm] are:
 [[ 1 55]
 [ 1 48]
 [ 1 88]
 [ 1 71]
 [ 2 11]]
 ```



## Problem 3
**3.1** In a separate .py file named `disp.py` , create a function called `displacement(v0, a, t)` that calculates the total displacememnt of a body during a time interval $t$, when the body has initial speed $v_0$ and a constant acceleration $a$. The function will return the displacement.

(The displacement of a body that moves with a constant acceleration will be $\Delta x = v_0 t + \frac{1}{2}at^2$)

Remember to start the function with a proper docstring.

- Import the function from the .py file. 

In [23]:
from disp import displacement

**3.2** Calculate the total displacement for:

$v_0 = 10 ~~\mathrm{m\cdot s^{-1}}$, $a = -9.8 ~~\mathrm{m\cdot s^{-2}}$,
and $t = 4 ~~\mathrm{s}$ 

- Print the result in an informative way.

In [24]:
v0=10
a=-9.8
t=4
dx = displacement(v0,a,t)
print("The displacement of a body with inital velocity of",v0,\
      "m/s undergoing a constant acceleration of",a,"m/s^2 for",t,"s is",dx,"m.")


The displacement of a body with inital velocity of 10 m/s undergoing a constant acceleration of -9.8 m/s^2 for 4 s is -38.400000000000006 m.


##### Expected Call/Output:
```
dx = displacement(10, -9.8, 4)

print("The displacement of a body with an initial velocity of",v0,\
      "m/s undergoing a constant acceleration of", a, "m/s^2 for", t, "s is", dx, "m.") 
```
`
The displacement of a body with an initial velocity of 10 m/s undergoing a constant acceleration of -9.8 m/s^2 for 4 s is -38.400000000000006 m.
`

**3.3**
- With *numpy arrays* it is not hard to calculate the displacement vector:

   $\Delta \vec{r} = \vec{r}_{f} - \vec{r}_{0} = \vec{v}_0 \cdot t + \frac{1}{2}\vec{a}\cdot t^2$

- Update your function to allow the user to input vectors for $\vec{v}_0$ and  $\vec{a}$ as numpy arrays.

  **Note: After modifying and saving your .py file, you will need to restart the kernel for your Jupyter Notebook. If you don't, it will not recognize the changes. You can restart your kernel from the 'Kernel' menu or by tapping 0 twice.** 
- Use your function to calculate the vector displacement of a body with initial velocity:

  $\vec{v}_0 = 3 \hat{i} + 1 \hat{j} - 4 \hat{k} ~~\mathrm{m\cdot s^{-1}}$

  undergoing constant acceleration:
  
  $\vec{a} = 2 \hat{i} - 7 \hat{j} - 1 \hat{k} ~~\mathrm{m\cdot s^{-2}}$

  for t = 3.3 s. Assign it to the variable `dr` and print it.

- Using numpy arrays it is also pretty easy to calculate the displacement vectors for many bodies with different (constant) accelerations and initial velocities. For example, the $i^{th}$ body would have a displacement given by:

    $\Delta \vec{r}_i = \vec{v}_{0i} \cdot t + \frac{1}{2}\vec{a}_i\cdot t^2$

  Think about how you would do that. Do you think you would need to modify your function?

In [28]:
v0 = np.array([3,1,-4]) #velocity in form xi,yj,zk
a = np.array([2,-7,-1]) #acceleration in form xi,yj,zk
t = 3.3 #s
dr = displacement(v0,a,t)
print(dr)

[ 20.79  -34.815 -18.645]


In [29]:
assert np.all(dr == [ 20.79,  -34.815, -18.645])


**3.4** Use your function from Problem 1 to determine the *time* it takes for object with constant acceleration $a$ and initial velocity $v_0$ to be displaced by $\Delta x$. Print an informative statement about the time `dt` it takes when $\Delta x = -20 ~\mathrm{m}$, $v_0 = +15 ~\mathrm{m\cdot s^{-1}}$ and $a = -9.8  ~\mathrm{m\cdot s^{-2}}$. Also store the time in seconds in the variable `dt`.

In [37]:
# displacement formula (1/2)*at^2 +v0t - d = 0
# squared value of t relate to a, t value relate to v0, and d value relate to x
a= -9.8 
displacement_accel = a *0.5
b= 15
c= -20
displacement_position = c *(-1)
dt = quadratic (displacement_accel,b,displacement_position)
print("Starting with inital velocity of",b,"m/s and a constant acceleration of",a,\
      "m/s^2 a body has a displacement of",c,"m after",dt[1],"s")


# ax^2 + bx + c  ax^2 = acceleration, bx = velocity c = position

Starting with inital velocity of 15 m/s and a constant acceleration of -9.8 m/s^2 a body has a displacement of -20 m after 4.065253540484535 s


##### Expected Output:
```
Starting with an initial velocity of 15.0  m/s and a constant acceleration of -9.8  m/s^2 a body has a displacement of -20.0  m after 4.065253540484535 s.
```



---
### Homework Submission
Prepare and submit your homework on JupyterHub.

- Make sure that all the cells run without errors and that your output matches the solution.

  *All your functions will be tested with different inputs to assure that they are working as intended!*
  
- Navigate to the `Home` tab in JupyterNotebook and make sure all additional files (.py files, figures, etc.) are in your assignment folder. 
- Copy your up-to-date `.log` file into your assignment folder.
- Make sure your submission folder is properly named according to the naming conventions.
- After submission, Feel free to download a copy of your submitted assignment folder.  Use GitHub to update and commit your changes on your machine and push to your online GitHub repository.
